File: src/Solution/DynamicPlan/划分型/分割回文串.java
package Solution.DynamicPlan.划分型;

import org.junit.Test;

/**
 * https://leetcode-cn.com/problems/palindrome-partitioning-ii/
 * 重复枚举最后一段，判断最后一段是不是回文串
 */
public class 分割回文串 {
    public int minCut(String s) {
        int[] dp = new int[s.length() + 1];
        boolean[][] isPalindrome = getIsPalindrome(s);
        dp[0] = -1;
        for(int i = 2; i <= s.length(); i++) {
            int min = Integer.MAX_VALUE;
            for(int j = 1; i - j >= 0; j++) {
                if(isPalindrome[i - j][i - 1] && min > dp[i - j]) {
                    min = dp[i - j];
                }
            }
            dp[i] = min + 1;
        }

        return dp[s.length()];
    }

    /**
     * 生成字符串所有回文子串的判断数组
     * @param str 字符串父串
     * @return 判断结果数组
     */
    private boolean[][] getIsPalindrome(String str) {
        boolean[][] isPalindrome = new boolean[str.length()][str.length()];

        for(int len = 1; len <= str.length(); len++) {
            for(int i = 0; i + len - 1 < str.length(); i++) {
                if(len == 1) isPalindrome[i][i] = true;
                else if(len == 2) isPalindrome[i][i + 1] = str.charAt(i) == str.charAt(i + 1);
                else {
                    int j = i + len - 1;
                    isPalindrome[i][j] = isPalindrome[i + 1][j - 1] && str.charAt(i) == str.charAt(j);
                }
            }
        }

        return isPalindrome;
    }

    @Test
    public void test() {
        String str = "aabcc";
        System.out.println(minCut(str));
    }
}

File: python/DP/划分型/分割回文串.py

def minCut(s):
    """
    将一个字符串分割为多个回文串，使得分割次数最少
    划分型DP问题，重复枚举最后一段找到结果
    https://leetcode-cn.com/problems/palindrome-partitioning-ii/
    LS：将字符串 s 前 i 个字符划分为回文串最少的划分次数为 x
        那么设最后一段长度为 k ，那么前 i-k 个字符划分为回文串最少次数为 x-1
        子问题出现了
    SQ：将字符串前
    MF：令 f(i) 为前 i 个字符划分为回文串的最少划分次数，b(i, j) 为判断以 i 起始 j 结尾的字符串是否为回文串
        f(i) = min{ f(i - k) } + 1, when b(i, j)
           1 <= k <= i - 1
    B：f(0) = 0
    :type s: str
    :rtype: int
    """
    is_pali = [[False for _ in range(len(s))] for _ in range(len(s))]
    

